File: scala/src/main/scala/scalasearch/SearchFile.scala
package scalasearch

import java.io.File

class SearchFile(val containers: List[String], val path: String, val file: String) {

  val CONTAINER_SEPARATOR = "!"

  def this(path: String, file: String) = {
    this(List.empty[String], path, file)
  }

  def toFile: File = {
    val p = new File(path)
    new File(p, file)
  }

  // get just the path inside the container(s)
  def getPath = {
    toFile.getPath
  }

  // get just the path inside the container(s)
  def getPathWithContainers = {
    toString
  }

  override def toString = {
    val sb = new StringBuilder
    if (containers.nonEmpty) {
      sb.append(containers.mkString(CONTAINER_SEPARATOR)).append(CONTAINER_SEPARATOR)
    }
    if (path.nonEmpty) {
      sb.append(path).append(File.separator)
    }
    sb.append(file)
    sb.toString()
  }
}

object SearchFileType extends Enumeration {
  type SearchFileType = Value
  val NonSearchFile = Value // a file that does not match search criteria
  val SearchFile = Value // a regular file that matches search criteria
  val ArchiveSearchFile = Value // an archive file that matches search criteria
}

File: java/src/main/java/javasearch/SearchFile.java
package javasearch;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class SearchFile {
    public static String CONTAINER_SEPARATOR = "!";
    private List<String> containers;
    private String path;
    private String fileName;
    private FileType fileType;

    public SearchFile(String path, String fileName, FileType fileType) {
        this(new ArrayList<String>(), path, fileName, fileType);
    }

    public SearchFile(List<String> containers, String path, String fileName,
                      FileType fileType) {
        this.containers = containers;
        this.path = path;
        this.fileName = fileName;
        this.fileType = fileType;
    }

    public void addContainer(String c) {
        containers.add(c);
    }

    public List<String> getContainers() {
        return this.containers;
    }

    protected void setContainers(List<String> cs) {
        this.containers = cs;
    }

    public String getPath() {
        return this.path;
    }

    protected void setPath(String p) {
        this.path = p;
    }

    public String getFileName() {
        return this.fileName;
    }

    protected void setFileName(String f) {
        this.fileName = f;
    }

    public FileType getFileType() {
        return this.fileType;
    }

    protected void setFileType(FileType ft) {
        this.fileType = ft;
    }

    public File toFile() {
        File dir = new File(path);
        return new File(dir, fileName);
    }

    public String toString() {
        StringBuilder sb = new StringBuilder();
        if (containers.size() > 0) {
            for (int i=0; i < containers.size(); i++) {
                if (i >0) sb.append(CONTAINER_SEPARATOR);
                sb.append(containers.get(i));
            }
            sb.append(CONTAINER_SEPARATOR);
        }
        if (path != null && !path.equals("")) {
            sb.append(path).append(File.separator);
        }
        sb.append(fileName);
        return sb.toString();
    }

}
